src/IQFeed.CSharpApiClient.Extensions/Lookup/Historical/Resample/TickMessageExtensions.cs

```diff
@@ -54,60 +55,70 @@ private static IEnumerable<HistoricalBar> ToHistoricalBarsAscending(this IEnumer
-                    if (tick.Last < currentBar.Low)
-                        currentBar.Low = tick.Last;
-
-                    if (tick.Last > currentBar.High)
-                        currentBar.High = tick.Last;
-
-                    currentBar.Close = tick.Last;
+                    // if we get here, we have a valid trade tick, and we need a new bar
+                    var currentTimestamp = tick.Timestamp.Trim(intervalTicks);
+                    currentBar = new HistoricalBar(
+                        currentTimestamp,
+                        tick.Last,
+                        tick.Last,
+                        tick.Last,
+                        tick.Last,
+                        0,
+                        0,
+                        0,
+                        0,
+                        tick.Last * tick.LastSize);
+                    nextTimestamp = currentTimestamp.AddTicks(intervalTicks);
+                    currentDate = currentTimestamp.Date;
+                }
 
-                    currentBar.PeriodVolume += tick.LastSize;
-                    currentBar.PeriodTrade += 1;
+                if (tick.Last < currentBar.Low)
+                    currentBar.Low = tick.Last;
 
-                    currentBar.TotalVolume = totalVolume;
-                    currentBar.TotalTrade = totalTrade;
+                if (tick.Last > currentBar.High)
+                    currentBar.High = tick.Last;
 
-                    currentBar.VWAP += tick.Last * tick.LastSize;
-                    continue;
-                }
-
-                if (currentBar != null)
-                {
-                    // reset the counts if dates differ
-                    if (tick.Timestamp.Date != currentBar.Timestamp.Date)
-                    {
-                        totalVolume = tick.LastSize;
-                        totalTrade = 1;
-                    }
+                currentBar.Close = tick.Last;
 
-                    currentBar.VWAP = currentBar.VWAP / currentBar.PeriodVolume;
-                    yield return currentBar;
-                }
+                currentBar.PeriodVolume += tick.LastSize;
+                currentBar.PeriodTrade += 1;
 
-                var currentTimestamp = tick.Timestamp.Trim(intervalTicks);
-                nextTimestamp = currentTimestamp.AddTicks(intervalTicks);
+                currentBar.TotalVolume = totalVolume;
+                currentBar.TotalTrade = totalTrade;
 
-                currentBar = new HistoricalBar(
-                    currentTimestamp,
-                    tick.Last,
-                    tick.Last,
-                    tick.Last,
-                    tick.Last,
-                    totalVolume,
-                    tick.LastSize,
-                    totalTrade,
-                    1,
-                    tick.Last * tick.LastSize);
+                currentBar.VWAP += tick.Last * tick.LastSize;
             }
 
             // return the last created bar when last tick reached
```

Comment:
does this code still relevant ?

```
  if (currentBar != null)
            {
                currentBar.VWAP = currentBar.VWAP / currentBar.PeriodVolume;
                yield return currentBar;
            }
```

Reply:
Yes, it's necessary to return the very last bar. The very last tick can only trigger the closing of the previous bar, so that's there outside the tick loop to close the final bar.

Reply:
okay just very last thing, can you confirm what I just sent by email?